### PR TITLE
[CARBONDATA-301] Added Sort processor step for data loading.

### PR DESCRIPTION
**Pull request by ravipesala**
Add SortProcessorStep which sorts the data as per dimension order and write the sorted files to temp location and merge sort it.

### CODE REVIEW

File: processing/src/main/java/org/apache/carbondata/processing/newflow/steps/sort/SortProcessorStepImpl.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+import org.apache.carbondata.processing.sortandgroupby.sortdata.SortParameters;
+import org.apache.carbondata.processing.store.SingleThreadFinalSortFilesMerger;
+import org.apache.carbondata.processing.store.writer.exception.CarbonDataWriterException;
+import org.apache.carbondata.processing.util.CarbonDataProcessorUtil;
+
+/**
+ * It sorts the data and write them to intermediate temp files. These files will be further read
+ * by next step for writing to carbondata files.
+ */
+public class SortProcessorStepImpl extends AbstractDataLoadProcessorStep {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/steps/sort/SortProcessorStepImpl.java]: In this step, better to abstract an interface called `Sorter`. We can move current sort implementation (SortDataRows/SortIntermediateFileMerger) into `MergeSortSorter` which implements `Sorter`, and we can add `ExternalSorter` later for in memory sort

So in this class, there will be an member variable called sorter.

Or, we can create another PR to do it if hard to do in one PR

File: processing/src/main/java/org/apache/carbondata/processing/newflow/sort/impl/CarbonParallelReadMergeSorterImpl.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+  public CarbonParallelReadMergeSorterImpl(DataField[] inputDataFields) {
+    this.inputDataFields = inputDataFields;
+  }
+
+  @Override
+  public void initialize(SortParameters sortParameters) {
+    this.sortParameters = sortParameters;
+    intermediateFileMerger = new SortIntermediateFileMerger(sortParameters);
+    String storeLocation = CarbonDataProcessorUtil
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/sort/impl/CarbonParallelReadMergeSorterImpl.java]: move `CarbonDataProcessorUtil` to next line and break line in parameters

File: processing/src/main/java/org/apache/carbondata/processing/newflow/sort/CarbonSorter.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import org.apache.carbondata.processing.newflow.exception.CarbonDataLoadingException;
+import org.apache.carbondata.processing.newflow.row.CarbonRowBatch;
+import org.apache.carbondata.processing.sortandgroupby.sortdata.SortParameters;
+
+/**
+ * This interface sorts all the data of iterators.
+ * The life cycle of this interface is initialize -> sort -> close
+ */
+public interface CarbonSorter {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/sort/CarbonSorter.java]: The sorter should have following interface
```
// add rows into the sorter, sorter can sort it on the fly
void insertRow(CarbonRowBatch)

// return an iterator that is sorted
Iterator getIterator()
```

File: processing/src/main/java/org/apache/carbondata/processing/newflow/steps/SortProcessorStepImpl.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      LogServiceFactory.getLogService(SortProcessorStepImpl.class.getName());
+
+  private CarbonSorter carbonSorter;
+
+  public SortProcessorStepImpl(CarbonDataLoadConfiguration configuration,
+      AbstractDataLoadProcessorStep child) {
+    super(configuration, child);
+  }
+
+  @Override public DataField[] getOutput() {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/steps/SortProcessorStepImpl.java]: move override to previous line

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java [hunk 5]
```diff
@@ -334,24 +151,24 @@ public void startSorting() throws CarbonSortKeyAndGroupByException {
       toSort = new Object[entryCount][];
       System.arraycopy(recordHolderList, 0, toSort, 0, entryCount);
 
-      if (noDictionaryCount > 0) {
-        Arrays.sort(toSort, new RowComparator(noDictionaryDimnesionColumn, noDictionaryCount));
+      if (parameters.getNoDictionaryCount() > 0) {
+        Arrays.sort(toSort, new RowComparator(parameters.getNoDictionaryDimnesionColumn(),
+            parameters.getNoDictionaryCount()));
       } else {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java]: remove empty line

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java [hunk 6]
```diff
@@ -334,24 +151,24 @@ public void startSorting() throws CarbonSortKeyAndGroupByException {
...
+        Arrays.sort(toSort, new RowComparatorForNormalDims(parameters.getDimColCount()));
       }
       recordHolderList = toSort;
 
       // create new file
-      File file =
-          new File(this.tempFileLocation + File.separator + this.tableName + System.nanoTime() +
-              CarbonCommonConstants.SORT_TEMP_FILE_EXT);
+      File file = new File(
+          parameters.getTempFileLocation() + File.separator + parameters.getTableName() + System
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java]: move System to next line

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdatastep/SortKeyStep.java [hunk 7]
```diff
@@ -171,15 +178,16 @@ else if (RemoveDictionaryUtil.checkAllValuesForNull(row)) {
 
       this.noDictionaryColMaping =
           RemoveDictionaryUtil.convertStringToBooleanArr(meta.getNoDictionaryDimsMapping());
-
-      this.sortDataRows = new SortDataRows(meta.getTabelName(),
-          meta.getDimensionCount() - meta.getComplexDimensionCount(),
-          meta.getComplexDimensionCount(), meta.getMeasureCount(), this.observer,
-          meta.getNoDictionaryCount(), meta.getPartitionID(), meta.getSegmentId() + "",
-          meta.getTaskNo(), this.noDictionaryColMaping);
+      SortParameters parameters = SortParameters
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdatastep/SortKeyStep.java]: move SortParameters to next line

File: processing/src/main/java/org/apache/carbondata/processing/util/RemoveDictionaryUtil.java [hunk 8]
```diff
@@ -123,6 +123,60 @@ private static int calculateTotalBytes(ByteBuffer[] byteBufferArr) {
   }
 
   /**
+   * This method will form one single byte [] for all the high card dims.
+   *
+   * @param byteBufferArr
+   * @return
+   */
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/util/RemoveDictionaryUtil.java]: Why this class is called RemoveDictionaryUtil?

File: processing/src/main/java/org/apache/carbondata/processing/newflow/sort/impl/CarbonParallelReadMergeSorterImpl.java (hunk 2 above)
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/sort/impl/CarbonParallelReadMergeSorterImpl.java]: I guess PR 217 is not merged and once it is merged those changes would be reflected here. And jira 287 would be sufficient for it.

File: processing/src/main/java/org/apache/carbondata/processing/newflow/steps/SortProcessorStepImpl.java (hunk 4 above)
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/steps/SortProcessorStepImpl.java]: ok

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java (hunk 5 above)
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java]: ok

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java (hunk 6 above)
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java]: ok

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdatastep/SortKeyStep.java (hunk 7 above)
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdatastep/SortKeyStep.java]: ok

File: processing/src/main/java/org/apache/carbondata/processing/util/RemoveDictionaryUtil.java (hunk 8 above)
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/util/RemoveDictionaryUtil.java]: It is named like this long time back. We can remove it at the time of removing kettle based code.

**Comment by jackylk** [comment]
code style check is failing
/Users/jackylk/code/incubator-carbondata/processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortIntermediateFileMerger.java:93: error: 'method call' child have incorrect indentation level 4, expected level should be 6.
/Users/jackylk/code/incubator-carbondata/processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortIntermediateFileMerger.java:93: error: 'try' child have incorrect indentation level 4, expected level should be 6.
/Users/jackylk/code/incubator-carbondata/processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortIntermediateFileMerger.java:94: error: 'method call' child have incorrect indentation level 4, expected level should be 6.
/Users/jackylk/code/incubator-carbondata/processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortIntermediateFileMerger.java:94: error: 'try' child have incorrect indentation level 4, expected level should be 6.
/Users/jackylk/code/incubator-carbondata/processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortIntermediateFileMerger.java:95: error: 'try rcurly' have incorrect indentation level 2, expected level should be 4.
/Users/jackylk/code/incubator-carbondata/processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortIntermediateFileMerger.java:96: error: 'catch' child have incorrect indentation level 4, expected level should be 6.
/Users/jackylk/code/incubator-carbondata/processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortIntermediateFileMerger.java:97: error: 'catch rcurly' have incorrect indentation level 2, expected level should be 4.

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**